scripts/kfctl.sh

```diff
@@ -8,6 +8,60 @@
 # cd myapp
 # kfctl.sh generate all
 # kfctl.sh apply all
+#
+# kfctl will assemble and deploy prototypes using ksonnet
+# Prototypes are grouped by component and are available
+# after installing their component
+#    `ks pkg install kubeflow/<component>`
+# Components can be grouped under modules by doing
+#    `ks module create <module-name>`
+#    `ks generate <prototype-name> <component-name> --module <module-name>`
+# One or more modules can be separately deployed by doing
+#    `ks env targets default --module <module-name> ...`
+#    `ks apply default`
+# ksonnet modules are used so that different combinations of prototypes 
+# can be grouped within a ksonnet environment. By setting the environment targets
+# as one or more modules the environment becomes composable either before deployment
+# or after. See `kfctl.sh add --help`.
+#
+# Syntax:
+# A module is named along with it's dependencies by doing
+#   `kfctl.sh add <module-name> --dependsOn <component-name> [<component-name>] ...`
+# More than one component will result in nested modules. 
+#
+# Use case 1: deploying a component's prototype along with it's dependencies
+#   The notebooks component can use protected namespaces so new notebooks are deployed
+#   within this namespace. In this case the notebook component has dependencies on
+#   the profiles and metacontroller components. In this case you would do
+#
+#   `kfctl.sh add notebooks --dependsOn notebooks profiles metacontroller`
+#
+# Use case 2: deploying some prototypes of a component
+#   The kubeflow/core component contains a number of prototypes which are specific to GKE 
+#   like cloud-endpoints, iap-ingress. Other prototypes in core like ambassador and 
+#   centraldashboard should be always be deployed. In this case you can do
+#
+#   `kfctl.sh add core --dependsOn 'core/{ambassador,centraldashboard}'`
+#
+#   In this case we are creating a module called core and adding the ambassador and 
+#   centraldashboard prototypes from the core component.
+# 
+# Use case 3: deploying a prototype in a component after the initial deployment
```

Comment:
both modules and components can be used across different environments but modules aren't added to an environment by default. Also modules can contain an arbitrary grouping of components so it would be easier for me to define something like 

`openvino --dependsOn openvino core/{ambassador,centraldashboard}`

as a self contained module which isn't included in an environment by default and has the minimal components to define an app.

Reply:
Short summary of modules is [here](https://github.com/ksonnet/ksonnet/blob/master/docs/concepts.md#module)